### ISSUE TITLE
Poor performance using test_any_model.py on ShapeNet Part segmentation

### ISSUE THREAD

**Issue by CZ-Wu** [HuguesTHOMAS/KPConv#117]
During training, I get mIoU=86% and mcIoU=85%, but it seems that you do not use all of data in test set. I use  `test_any_model.py`  to test the trained model at epoch 500, but I get poor performance (mIoU=40.7% and mcIoU=34.3%). How to fix this issue?

**Comment by HuguesTHOMAS** [comment]
Are you sure the shapes are oriented the same in the training and test set?

**Comment by CZ-Wu** [comment]
> Are you sure the shapes are oriented the same in the training and test set?

Yes. The code in `test_any_model.py` read the pickle file and `.ply` files generated by your training code.

**Comment by HuguesTHOMAS** [comment]
The reason has to be some kind of bug somewhere. Given the very large discrepancy between your validation and test scores, it should be some problem with the pointclouds themselves. 

I think you should be able to see the problem if you save the pointcloud used by the network in both cases and check them with a visualization tool like CloudCompare. In the test script, save the points and label at this point:
https://github.com/HuguesTHOMAS/KPConv/blob/07861d8e40244c20ede5a362052e01f61dcc012a/utils/tester.py#L302-L307

And at the similar point in the validation script.

You can use the write_ply function provided.

**Comment by CZ-Wu** [comment]
> The reason has to be some kind of bug somewhere. Given the very large discrepancy between your validation and test scores, it should be some problem with the pointclouds themselves.
> 
> I think you should be able to see the problem if you save the pointcloud used by the network in both cases and check them with a visualization tool like CloudCompare. In the test script, save the points and label at this point:
> https://github.com/HuguesTHOMAS/KPConv/blob/07861d8e40244c20ede5a362052e01f61dcc012a/utils/tester.py#L302-L307
> 
> And at the similar point in the validation script.
> 
> You can use the write_ply function provided.

Thanks a lot. I find that you have changed the order of 'xyz' to 'xzy' when you read the `.ply` file:
https://github.com/HuguesTHOMAS/KPConv/blob/07861d8e40244c20ede5a362052e01f61dcc012a/utils/tester.py#L443-L445
I get the normal results after modifying the code: mIoU=86.1% and mcIoU=84.2%. But the reported results in your paper are mIoU=86.4% and mcIoU=85.1%, there are gaps, especially for mcIoU. I use the default params to train the model, how can I get the results in your paper?

**Comment by HuguesTHOMAS** [comment]
The paper results were obtained with the Tensorflow implementation and I am not sure if I kept the same exact parameters (expecially the network architecture and number of parameters). Some things also changed from one implementation to the other (for example how we calibrate the pointcloud sizes and neighborhoods etc), so it is not even sure that you can reproduce the same performances with PyTorch.  

That being said, you should be able to get better mcIoU by changing some of the parameters, using a bigger network for example, or playing with the learning rate and its decay for a better convergence.

**Comment by CZ-Wu** [comment]
I really can not reproduce the results in your paper. Did you save the checkpoint?

**Comment by HuguesTHOMAS** [comment]
I conducted the experiments a long time ago and I did not keep the logs, I am sorry that I can't help you. There are so many parameters that I can't tell you for sure what they were when I got this best result.

**Comment by densechen** [comment]
> > The reason has to be some kind of bug somewhere. Given the very large discrepancy between your validation and test scores, it should be some problem with the pointclouds themselves.
> > I think you should be able to see the problem if you save the pointcloud used by the network in both cases and check them with a visualization tool like CloudCompare. In the test script, save the points and label at this point:
> > https://github.com/HuguesTHOMAS/KPConv/blob/07861d8e40244c20ede5a362052e01f61dcc012a/utils/tester.py#L302-L307
> > 
> > And at the similar point in the validation script.
> > You can use the write_ply function provided.
> 
> Thanks a lot. I find that you have changed the order of 'xyz' to 'xzy' when you read the `.ply` file:
> https://github.com/HuguesTHOMAS/KPConv/blob/07861d8e40244c20ede5a362052e01f61dcc012a/utils/tester.py#L443-L445
> 
> 
> I get the normal results after modifying the code: mIoU=86.1% and mcIoU=84.2%. But the reported results in your paper are mIoU=86.4% and mcIoU=85.1%, there are gaps, especially for mcIoU. I use the default params to train the model, how can I get the results in your paper?

Hi, @CZ-Wu , I also occured the same problem with you. Is it the following order?
```
points = np.vstack((data['x'], -data['y'], data['z'])).T 
```
